problem_builder/tasks.py

```diff
@@ -108,4 +106,60 @@ def scan_for_blocks(block):
         "report_filename": filename,
         "start_timestamp": start_timestamp,
         "generation_time_s": generation_time_s,
+        "display_data": [] if len(rows) == 1 else rows[1:]
     }
+
+
+def _extract_data_for_display(submission, student_id, block_type, match_string):
+    """
+    Extract data that will be displayed on Student Answers Dashboard
+    from `submission`.
+    """
+
+    # Username
+    user = user_by_anonymous_id(student_id)
+    username = user.username
+
+    # Question
+    student_item = StudentItem.objects.get(pk=submission['student_item'])
+
+    block_key = UsageKey.from_string(student_item.item_id)
+    block = modulestore().get_item(block_key)
+
+    # Answer
+    answer = submission['answer']
+
+    try:
+        choices = block.children
+    except AttributeError:
+        pass
+    else:
+        for choice in choices:
+            choice_block = modulestore().get_item(choice)
+            if choice_block.value == answer:
+                answer = choice_block.content
+                break
+
+    # Short-circuit if answer does not match search criteria
+    if not match_string.lower() in answer.lower():
+        return
+
+    # Unit
+    mentoring_block = modulestore().get_item(block.parent)
+    unit = modulestore().get_item(mentoring_block.parent)
+
+    # Subsection
+    subsection = modulestore().get_item(unit.parent)
+
+    # Section
+    section = modulestore().get_item(subsection.parent)
```

Comment:
A couple of notes here:

1. This method is called for each block and for each student. So if you have one problem block that has 10,000 students, this method will be called 10,000 times. But the `mentoring_block`, `unit`, `subsection`, `section` data is the same for each block. Since the `get_item()` call is expensive, you should move all of these lookups to be outside of the loop, so they are only queried once per block, and not once per block-student pair.

2. The unit is not necessarily the direct parent of the problem block. Configurations like the following are possible:

    ```
Course > section (chapter) > subsection (sequential) > unit (vertical) > A/B Test > Problem Builder > [MCQ/Answer/etc.]

Course > section (chapter) > subsection (sequential) > unit (vertical) > vertical > vertical > Library Content Block > vertical > Problem Builder > [MCQ/Answer/etc.]
    ```
    A more reliable way to code this would be:
   
    ```python
    block_names_by_type = {}
    block_iter = block
    while block_iter:
        block_iter_type = block_iter.scope_ids.block_type
        block_names_by_type[block_iter_type] = block.display_name_with_default
        block_iter = block_iter.get_parent() if block_iter.parent else None
    section_name = block_names_by_type.get('chapter', '')
    subsection_name = block_names_by_type.get('sequential', '')
    unit_name = block_names_by_type.get('vertical', '')
    ```
   (and do that outside of this method)